docs/design/cpsubsystem/04-cp-object-lifecycle.md

```diff
@@ -0,0 +1,271 @@
+
+For tracking CP object lifecycle (creation/destruction) each internal CP object should have a unique object ID (UUID):
+
+- A Hazelcast client should obtain an object ID during the Client-proxy creation
+- Each Client → Server CP method call (operation) should include object ID
+- If there is no such object ID on the Server side, there should be an exception raised and the Client has to recreate the proxy to obtain a new object ID
+- When the CP object is destroyed, the object ID should be destroyed
+- The object ID should be persisted and maintained in a consistent way (the same as an object value) on all CP group members
+
+To implement a such behaviour a new codec, message and Raft operation were added:
+- [CPGroupCreateCPObjectCodec](../../../hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/CPGroupCreateCPObjectCodec.java) 
+- [CreateRaftObjectMessageTask](../../../hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/client/CreateRaftObjectMessageTask.java)
+- [CreateOrGetRaftObjectOp](../../../hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/operation/CreateOrGetRaftObjectOp.java)
+
+This flow creates an empty data structure of a specific type (serviceName) on the Server side with an UUID and returns this UUID to the Client side:
+
+```java
+private UUID getObjectId(RaftGroupId groupId, String serviceName, String objectName) {
+    ClientMessage request = CPGroupCreateCPObjectCodec.encodeRequest(groupId, serviceName, objectName);
+    ClientMessage response = new ClientInvocation(client, request, objectName).invoke().joinInternal();
+    return CPGroupCreateCPObjectCodec.decodeResponse(response);
+}
+```
+
+This UUID is recorded as a property of the object proxy on the Client side:
+
+```java
+public class AtomicLongProxy extends ClientProxy implements IAtomicLong {
+
+    private final RaftGroupId groupId;
+    private final String objectName;
+    private UUID objectUUID;
+
+    public AtomicLongProxy(ClientContext context, RaftGroupId groupId, String proxyName, String objectName, UUID objectUUID) {
+    ...
+    }
+}
+```
+
+On the Server side, each CP data structure (at the current moment `AtomicLong` and `AtomicRef`) should also have a UUID property:
+
+```java
+public abstract class RaftAtomicValue<T> {
+
+    private final CPGroupId groupId;
+    private final String name;
+    private UUID uuid;
+
+    public RaftAtomicValue(CPGroupId groupId, String name, UUID uuid) {
+        this.groupId = groupId;
+        this.name = name;
+        this.uuid = uuid;
+    }
+    ...
+}
+```
+
+To minimize the number of changes on a client/server side, and do not recreate/update dozens of codecs, messages and Raft operations, object UUID and object name are concatenated on the Client side. This combined object name passes via existing API during each operation call:
```

Comment:
Can you make a comparison of the two approaches (new client messages for object creation+name mangling vs extending existing client messages & ops with a nullable `id` field)?
The need to update several existing messages & operations is clearly a disadvantage for the latter. Any advantages to that approach?
Indicative themes to consider when comparing the alternatives (maybe not all make sense for comparison, these are just suggestions): source maintainability, runtime compatibility (client-to-member and member-to-member), persistent log compatibility (can I start a new member that operates with object IDs off a CP persistence directory from 5.3 member?) , number of message exchanges/network overhead.

Reply:
Will make a comparison